openstackmap.py

```diff
@@ -24,245 +24,480 @@
-        users = list(user.to_dict() for user in keystone.users.list())
-        pickle.dump(users, f, protocol=-1)
-
-    with open(pers_keystone + '/tenants.pickle', 'wb') as f:
-        tenants = list(tenant.to_dict() for tenant in keystone.projects.list())
-        pickle.dump(tenants, f, protocol=-1)
-
-    with open(pers_keystone + '/asignments.pickle', 'wb') as f:
-        a = list(asig.to_dict() for asig in keystone.role_assignments.list())
-        roles_a = a
-        pickle.dump(roles_a, f, protocol=-1)
-
-def load_keystone():
-    global users_by_id, users_by_email, tenants_by_id, tenants_by_name
-    global roles_by_user, roles_by_project
-    global users, roles, tenants, roles_a 
-
-    if not os.path.exists(pers_keystone + '/users.pickle'):
-        refresh_keystone()
-    else: 
-        users = load_fkeystone('users')
-        tenants = load_fkeystone('tenants')
-        roles_a = load_fkeystone('asignments')
-        roles = load_fkeystone('roles')
-
-    users_by_email = dict()
-    users_by_id = dict()
-    for user in users:
-        if 'name' in user:
-            users_by_email[user['name']] = user
-        users_by_id[user['id']] = user
-
-    tenants_by_name = dict()
-    tenants_by_id = dict()
-    for tenant in tenants:
-        tenants_by_id[tenant['id']] = tenant
-        tenants_by_name[tenant['name']] = tenant
-
-    rolesdict = dict((role['id'], role['name']) for role in roles)
-    roles_by_user = dict()
-    roles_by_project = dict()
-    for roleasig in roles_a:
-        userid = roleasig['user']['id']
-        if 'project' in roleasig['scope']:
-            projectid = roleasig['scope']['project']['id']
+        if region:
+            self.osclients.set_region(region)
         else:
-            projectid = str(roleasig['scope'])
-        roleid = roleasig['role']['id']
-        if userid not in roles_by_user:
-            roles_by_user[userid] = list()
-        if projectid not in roles_by_project:
-            roles_by_project[projectid] = list()
-        roles_by_user[userid].append((rolesdict[roleid], projectid))
-        roles_by_project[projectid].append((rolesdict[roleid], userid))
-
-def refresh_neutron():
-    neutron = osclients.get_neutronclient()
-    global networks, subnets, routers, floatingips 
-    global security_groups, ports
-
-    with open(pers_region + '/networks.pickle', 'wb') as f:
-        networks = neutron.list_networks()['networks']
-        pickle.dump(networks, f, protocol=-1)
+            if 'OS_REGION_NAME' not in env:
+                raise Exception('Region parameter must be provided or '
+                                'OS_REGION_NAME variable must be defined.')
+            else:
+                region = env['OS_REGION_NAME']
+
+        self.persistence_dir = os.path.expanduser(persistence_dir)
+        self.pers_region = self.persistence_dir + '/' + region
+        self.pers_keystone = self.persistence_dir + '/keystone'
+        # Keystone resources
```

Comment:
I think that the operations to create the dict() of the self.* variables could be manage outside of this method in order to understand better the code and reduce the number of code lines in this method.